homeassistant/components/directv/config_flow.py

```diff
@@ -0,0 +1,130 @@
+"""Config flow for DirecTV."""
+import logging
+from typing import Any, Dict, Optional
+from urllib.parse import urlparse
+
+from DirectPy import DIRECTV
+from requests.exceptions import RequestException
+import voluptuous as vol
+
+from homeassistant.components.ssdp import ATTR_SSDP_LOCATION, ATTR_UPNP_SERIAL
+from homeassistant.config_entries import CONN_CLASS_LOCAL_POLL, ConfigFlow
+from homeassistant.const import CONF_HOST, CONF_NAME
+from homeassistant.core import callback
+from homeassistant.exceptions import HomeAssistantError
+
+from .const import DEFAULT_PORT
+from .const import DOMAIN  # pylint: disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+ERROR_CANNOT_CONNECT = "cannot_connect"
+ERROR_UNKNOWN = "unknown"
+
+DATA_SCHEMA = vol.Schema({vol.Required(CONF_HOST): str})
+
+
+def validate_input(data: Dict) -> Dict:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+    # directpy does IO in constructor.
+    dtv = DIRECTV(data["host"], DEFAULT_PORT)
+    version_info = dtv.get_version()
+
+    return {
+        "title": data["host"],
+        "host": data["host"],
+        "receiver_id": "".join(version_info["receiverId"].split()),
+    }
+
+
+class DirecTVConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for DirecTV."""
+
+    VERSION = 1
+    CONNECTION_CLASS = CONN_CLASS_LOCAL_POLL
+
+    @callback
+    def _show_form(self, errors: Optional[Dict] = None) -> Dict[str, Any]:
+        """Show the form to the user."""
+        return self.async_show_form(
+            step_id="user", data_schema=DATA_SCHEMA, errors=errors or {},
+        )
+
+    async def async_step_import(
+        self, user_input: Optional[Dict] = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initialized by yaml file."""
+        return await self.async_step_user(user_input)
+
+    async def async_step_user(
+        self, user_input: Optional[Dict] = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initialized by user."""
+        if not user_input:
+            return self._show_form()
+
+        errors = {}
+
+        try:
+            info = await self.hass.async_add_executor_job(validate_input, user_input)
+            user_input[CONF_HOST] = info[CONF_HOST]
+        except RequestException:
+            errors["base"] = ERROR_CANNOT_CONNECT
+            return self._show_form(errors)
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
+            errors["base"] = ERROR_UNKNOWN
+            return self._show_form(errors)
```

Comment:
What should the user do in this case? Normally we abort immediately if the user cannot amend the situation.

Reply:
valid point, it could of been the components I referenced as a start that had this. still new to python/ha dev in general but I'll continue doing some refinement during beta